src/main/java/org/icatproject/core/manager/search/OpensearchScriptBuilder.java

```diff
@@ -0,0 +1,210 @@
+package org.icatproject.core.manager.search;
+
+import java.util.Set;
+
+import javax.json.Json;
+import javax.json.JsonObjectBuilder;
+
+public class OpensearchScriptBuilder {
```

Comment:
The short answer is that ES/OS allow you to specify more complex logic via scripts: https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-scripting.html
In our case, I'm using these handle updates to nested objects. Relationships between entities are not suited to inverted indices, and so (for example) updating the `numericValue` on a specific `InvestigationParameter` is hard because all the `InvestigationParameters` are actually just in arrays on `Investigation` documents. Happy to talk about it more depth as needed.